torch_np/_decorators.py

```diff
@@ -0,0 +1,112 @@
+        if dtype is not None:
+            dtype = _dtypes.dtype(dtype)
+            torch_dtype = dtype._scalar_type.torch_dtype
+        return func(*args, dtype=torch_dtype, **kwds)
+    return wrapped
+
+
+def emulate_out_arg(func):
+    """Simulate the out=... handling *for functions which do not need it*.
+
+    With this decorator, the inner function just does not see the out array.
+    """
+    def wrapped(*args, out=None, **kwds):
+        from ._ndarray import ndarray
+        if out is not None:
+            if not isinstance(out, ndarray):
+                raise TypeError("Return arrays must be of ArrayType")        
+        result_tensor = func(*args, **kwds)
+        return _helpers.result_or_out(result_tensor, out)
+
+    return wrapped
+
+
+def out_shape_dtype(func):
+    """Handle out=... kwarg for ufuncs.
+
+    With ufuncs, `out` array can typcast and broadcast ufunc arguments, hence
+    extract the shape and dtype of the tensor which backs the `out` array
+    and pass these through.
+    """
+    def wrapped(*args, out=None, **kwds):
+        from ._ndarray import ndarray
+        if out is not None:
+            if not isinstance(out, ndarray):
+                raise TypeError("Return arrays must be of ArrayType")
+            kwds.update({'out_shape_dtype': (out.get().dtype, out.get().shape)})
+        result_tensor = func(*args, **kwds)
+        return _helpers.result_or_out(result_tensor, out)
+
+    return wrapped
+
+
+def deco_unary_ufunc_from_impl(impl_func):
+    @functools.wraps(impl_func)
+    @dtype_to_torch
+    @out_shape_dtype
+    def wrapped(x1, *args, **kwds):
+        from ._ndarray import asarray
+        x1_tensor = asarray(x1).get()
+        result = impl_func((x1_tensor,), *args, **kwds)
+        return result
+    return wrapped
+
+
+# TODO: deduplicate with _ndarray/asarray_replacer,
+# and _wrapper/concatenate et al
+def deco_ufunc_from_impl(impl_func):
+    @functools.wraps(impl_func)
+    @dtype_to_torch
+    @out_shape_dtype
+    def wrapped(x1, x2, *args, **kwds):
+        from ._ndarray import asarray
+        x1_tensor = asarray(x1).get()
+        x2_tensor = asarray(x2).get()
+        return impl_func((x1_tensor, x2_tensor), *args, **kwds)
+    return wrapped
+
+
+
+def axis_keepdims_wrapper(func):
+    """`func` accepts an array-like as a 1st arg, returns a tensor.
+
+    This decorator implements the generic handling of axis, out and keepdims
+    arguments for reduction functions.
+
+    Note that we peel off `out=...` and `keepdims=...` args (torch functions never
+    see them). The `axis` argument we normalize and pass through to pytorch functions.
+
+    """
+    # XXX: move this out of _ndarray.py (circular imports)
+    #
+    # TODO: 1. get rid of _helpers.result_or_out
+    #       2. sort out function signatures: how they flow through all decorators etc
+    @functools.wraps(func)
+    def wrapped(a, axis=None, out=None, keepdims=NoValue, *args, **kwds):
```

Comment:
I get the feeling that this function does too much. I really liked how you split the dtype and out processing above, but here you have mixed them all together. As a result, you seem to forget handling the `out` kwarg.

Reply:
In fact, `out=` arg should not be here already (removed in the last commit). 
Note the usage is `emulate_out_arg(axis_keepdims(...))`: https://github.com/Quansight-Labs/numpy_pytorch_interop/blob/refactor/torch_np/_ndarray.py#L281

This decorator itself simply unwraps ndarrays and passes heavy lifting to `_util.axis_keepdims` which does just what it says on the tin: handles axis tuples and `keepdims=True`.

Not sure how to simplify it further. Would a comment help?

The fact that the usage is not very clear is I guess a direct consequence of having separate decorators for various arguments. So if it's the direction we want to go, there will be more of this I'm afraid.

Reply:
So, there are two general issues here.
I don't see why would you always need to use this in conjunction to `emulate_out_arg`. These two functions may be used independently, so having `out` in the signature of this wrapper is unnecessary.
Even more, not only is unnecessary, but it's also incorrect. If you use it with `emulate_out_arg` afterwards, you happen to get the right behaviour almost by chance. Now, if you swap the order, you'll get a function with the same signature, but that takes an `out=` kwarg and... discards it!

As mentioned, I think that this function and the one that implements the out behaviour do two very different things, and should be independent of eachother.